chestnut-headphones: Add onboarding banner to homepage and profile page

taravancil opened the pull request:
## Links
* [chestnut-headphones.glitch.me/](https://chestnut-headphones.glitch.me/)
* [Onboarding design spec](https://docs.google.com/document/d/1iiJnU8COhu0EreV9i5xHaBMJRgeclaxLw6liBz-CNqU/)

![welcome](https://user-images.githubusercontent.com/7584833/63360502-7b900180-c33c-11e9-9024-2eca465489c5.gif)

## GIF/Screenshots:
<table>
<tr>
     <td><strong>large viewport</strong></td>
     <td><img src="https://user-images.githubusercontent.com/7584833/63358952-e12ebe80-c339-11e9-8105-05287cd31ca3.png"/>
  </td>
  </tr>
<tr>
     <td><strong>medium viewport</strong></td>
     <td><img src="https://user-images.githubusercontent.com/7584833/63358905-c8bea400-c339-11e9-8d9f-dbf2c531fbbd.png" />
  </td>
  </tr>
  <tr>
     <td><strong>small viewport</strong></td>
     <td><img height="550" src="https://user-images.githubusercontent.com/7584833/63358764-92812480-c339-11e9-96a6-6e778edc57e3.png"/></td>
  </tr>
</table>

## Changes:
* Add OnboardingBanner component
* Add CategoriesGrid component (used in OnboardingBanner and /create)
* Move shared constants from /create and the homepage to constants.styl
* Move NewProjectPop out of /header
* Update NewProjectPop to accept props for button styles and button text
* Update NewProjectPop analytics to send user id and the source of the clicks, i.e., navbar, onboarding banner, etc.

## How To Test:

Log in with a test account that doesn't have any projects
- You should see the onboarding banner on the homepage and on your profile page
- Change your screen width, the banner should look good
- Use the accessibility tools to inspect the welcome illustration. The label should be "Welcome"
- Click the "Create New Project" button, it should work.
- Click all the links in the onboarding banner, they should work.

Create a project
- You should not see the onboarding banner

Log out
- You should not see the onboarding banner

Various things to check
- The categories grid on /create still looks good
- The "New Project" button in the navbar still works

## Feedback I'm looking for:
- @ttseng does the banner look good to you?
- @sarahzinger does the data I'm sending along in the new project pop look good?

## Before deploying
- [ ] Run accessibility insights

ttseng commented:
nice!  the responsiveness works really well.  

the only change I'd suggest is adding a `margin-bottom: 7px` to `categoriesGridItem wrapItems img` so the category name isn't as close to the image:

![image](https://user-images.githubusercontent.com/519336/63361957-0ffb6380-c33f-11e9-88c4-0553835f6735.png)

sarahzinger commented on src/components/categories-grid/index.js:
```diff
@@ -0,0 +1,99 @@
+  const categoriesToRender = useMemo(
+    () => (categories !== allCategories ? allCategories.filter((category) => categories.includes(category.path)) : allCategories),
+    [categories],
+  );
```
what does this do?

taravancil replied:
This lets us optionally render only *some* of the categories. There are 8 total, but for example, the [grid in the onboarding banner](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R59) only shows 6/8: 'games', 'music', 'art', 'handy-bots', 'learn-to-code', 'tools-for-work'

sarahzinger replied:
cooool

sarahzinger commented on src/components/categories-grid/index.js:
```diff
@@ -0,0 +1,99 @@
+  return (
+    <ul className={classNames(styles.categoriesGrid, className)}>
+      {categoriesToRender.map((category) => (
+        <li key={category.path} className={itemClassNames} style={{ '--bg-color': category.color }}>
```
why `--bg-color` and not background-color?

taravancil replied:
🤷‍♀ idk tbh, you prefer `--background-color`?

sarahzinger replied:
i guess why does it have the `--` before it? is it deprecated? I thought you could do `style={{backgroundColor: category.color}}`

but maybe it doesn't matter 🤷‍♀

taravancil replied:
Oh I see, I'm setting a CSS variable - that's why it has the `--` in front. I could do it directly inline, but I'm not sure it matters that much

sarahzinger commented on src/components/constants.styl:
```diff
@@ -44,7 +45,27 @@ mediumSmallViewport = 592px
 
 side-padding = 20px
 
+u0 = 2px
```
so i recognize a lot of this is just moving around, but what is all this? can we put a comment for it? like what does u and t stand for?

taravancil replied:
I honestly do not know. @modernserf introduced these with the new homepage

modernserf replied:
The "u" units are margin & padding units; the "t"s are font sizes. Those all have clearer names in the shared component library

sarahzinger commented on src/components/new-project-pop/index.js:
```diff
@@ -93,16 +95,30 @@ const useNewProjectAPI = createAPIHook(async (api) => {
-    <PopoverWithButton onOpen={onOpen} buttonProps={{ size: 'small' }} buttonText="New Project">
-      {() => <NewProjectPop projects={projects} />}
+    <PopoverWithButton onOpen={onOpen} buttonProps={{ size: 'small', type: buttonType }} buttonText={buttonText}>
+      {() => <NewProjectPop projects={projects} align={align} analyticsProps={{ userId: currentUser.id || null, source }} />}
```
so this is totes chill but something we can do besides passing along props like this is to use AnalyticsContext instead if you're interested:
https://github.com/FogCreek/Glitch-Community/blob/4497ae553e1b37e3fee4268581ab90ec80329e1b/src/state/segment-analytics.js#L18
so like if this is on the user page for example you could pass currentUser.id at a higher level here and any analytics call on the userpage will have that 
https://github.com/FogCreek/Glitch-Community/blob/97eec7d4339492f103fab8f19cea8e8f31782ff1/src/presenters/pages/user.js#L217
and if you need to build on that origin in the function where you track things itself you can follow this inherited thing we do here:
https://github.com/FogCreek/Glitch-Community/blob/d5675ed420651b05f2c0dc13b2e46cd6b504095a/src/components/collection/create-collection-pop.js#L160

sarahzinger replied:
oh just realized you're also using https://github.com/FogCreek/Glitch-Community/blob/4497ae553e1b37e3fee4268581ab90ec80329e1b/src/state/segment-analytics.js#L68 but I think that will work the same way, we can double check together using segment if you want!

taravancil replied:
Oh sick, I'm making these updates now. Maybe if we have time later we can double check everything in Segment

sarahzinger commented on src/components/onboarding-banner/illustration.js:
```diff
@@ -0,0 +1,383 @@
+    <defs>
+      <path
+        d="M4.233.876c-2.53.959-3.958 3.376-3.192 5.402C1.81 8.3 4.482 9.164 7.013 8.204c2.527-.959 3.956-3.377 3.19-5.4C9.646 1.34 8.095.483 6.324.483a5.91 5.91 0 0 0-2.092.392z"
+        id="path-1"
```
i think we want to remove all ids from our inline svgs just incase they are ever reused on a page.

sarahzinger commented on src/components/onboarding-banner/index.js:
```diff
@@ -0,0 +1,81 @@
+        <Illustration />
+      </div>
+
+      <div className={styles.actions} style={{ background: isHomepage ? 'rgba(255, 255, 255, .875)' : 'none' }}>
```
why not do this with css?

taravancil replied:
No good reason really, other than I think I wrote this at 2am 🤦🏻‍♀️

sarahzinger commented on src/components/onboarding-banner/illustration.styl:
```diff
@@ -0,0 +1,7 @@
+
+@keyframes wave
+  33%
+    transform: rotate(-7deg);
```
I think you'll want to add a media query to not animate if the user prefers no motion https://github.com/FogCreek/Glitch-Community/blob/b33e7f7909495f496c89e34735de5693294eb8cc/src/presenters/pages/home-v2/animations.styl#L346

clottman commented on src/components/new-project-pop/index.js:
```diff
@@ -93,16 +95,30 @@ const useNewProjectAPI = createAPIHook(async (api) => {
   );
 }
 
+NewProjectPopButton.propTypes = {
```
Looks like we're missing prop types here - buttonType and source.

taravancil replied:
Added `source`, but invalid `buttonType`s will get caught in `Button` itself. Do you think it's worth duplicating for the sake of documentation?

clottman commented on src/components/onboarding-banner/illustration.styl:
```diff
@@ -0,0 +1,7 @@
+  animation: wave cubic-bezier(0.24, 0.45, 1, 1) 2s infinite forwards
+  transform-origin: 60px 121px
+
+@keyframes wave
```
Can we add in a media query to honor the user's `prefers-reduced-motion` setting, if set?

taravancil commented:
This is waiting of the infra team to whitelist the `ab-tests` cookie. They have a PR open, so I should be able to pick this back up after the holiday.

ttseng commented:
@keithk re: the links on the create page, are you referring to the categories at the bottom (rather than the framework or platform starters)?

![image](https://user-images.githubusercontent.com/519336/63974272-86ecd680-ca7a-11e9-9a9a-8064eee028fd.png)

(also, is that a concern for a separate ticket)?

taravancil commented:
A few notes on the changes I've pushed since this was last reviewed:

**A/B testing**
A/B testing is implemented! We're using a 50/50 onboard/don't onboard split. Pretty simple.

- [shared/ab-tests.js](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-58dbba706ff4b69f9bbd8ce359f855ea)
- [onboarding-banner/index.js#L24](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R24)

**AnalyticsContext**
Per Sarah's suggestion, I moved all of the custom analytics props into `<AnalyticsContext>`

- [onboarding-banner/index.js#L53](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R53)
- [header/index.js#L34](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-5c2ad23269ed1bd2a33c79c14b77e315R34)

**Shared components**
I pulled latest, which pulls in a few shared components. I updated the one emoji in the onboarding banner to use the `<Icon>` component. The shared-components library doesn't include `backhandIndex`, so until [this PR](https://github.com/FogCreek/shared-components/pull/45/files) is deployed, we're blocked

- [onboarding-banner/index.js#L85](https://github.com/FogCreek/Glitch-Community/pull/836/files#diff-ad7cb9c5b67e9f30910fcf625f867aa1R85)

taravancil commented:
Here's what a click on the "New Project" button in the onboarding banner will look like in Segment. It gives you `properties.origin`, which indicates whether the "New Project" button was clicked in the header or the onboarding banner.

Anything else we'll need to include?

```
{
  "anonymousId": "1234",
  "context": {
    "groupId": "0",
    "ip": "172.110.88.61",
    "library": {
      "name": "analytics.js",
      "version": "3.9.0"
    },
    "page": {
      "path": "/",
      "referrer": "",
      "search": "",
      "title": "Glitch",
      "url": "https://chestnut-headphones.glitch.me/"
    },
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:70.0) Gecko/20100101 Firefox/70.0"
  },
  "event": "open new-project pop",
  "integrations": {},
  "messageId": "ajs-e73772492f6bfeb946c0df7059517eb1",
  "originalTimestamp": "2019-09-10T14:38:49.258Z",
  "properties": {
    "origin": "homepage onboarding banner"
  },
  "receivedAt": "2019-09-10T14:38:49.305Z",
  "sentAt": "2019-09-10T14:38:49.260Z",
  "timestamp": "2019-09-10T14:38:49.303Z",
  "type": "track",
  "userId": "aUserId"
}
```

sarahzinger commented:
i'm currently seeing the onboarding banner on my user page but not on the homepage, is that a desired state?

GregWeil commented:
psa: use `document.cookie = 'ab-tests='+JSON.stringify({...AB_TESTS, Onboarding:'banner'})` to force yourself into the right a/b testing group

taravancil commented:
> i'm currently seeing the onboarding banner on my user page but not on the homepage

🤦‍♀ lmao no, i had toggled a flag for testing and never switched it back. @sarahzinger should be fixed now

Also note that you need to use a test account that has no projects to see the onboarding banner

ttseng commented:
Is it just me or the CTA styling for the `Create New Project` button not being applied?

![image](https://user-images.githubusercontent.com/519336/64644692-344bdc80-d3e1-11e9-8d96-f3f8e5e649d5.png)

(should be teal).

Otherwise, it looks great!

taravancil commented:
Yep, thanks for the catch Tiff! Must've gotten lost when I pulled in the styled components changes.

ttseng commented:
![wave](https://user-images.githubusercontent.com/519336/64712609-834d4c80-d489-11e9-8a86-59abd26f1450.gif)

Fixed hand direction + adjusted animation for computer illustration

taravancil commented:
Bumped the shared-compononents lib and added a fix to prevent the onboarding banner from being rendered on the server for users who shouldn't see it

sarahzinger commented:
still seems good to me :)
